### PR TITLE
fix compile warning add FreeTap() declaration and a conditional compiling, fix libssl-dev package name, update change log

### PR DESCRIPTION
**Pull request by youxiaojie**
fix compiling warning "implicit declaration" of Console.c and BridgeUnix.c

Changes proposed in this pull request:
  

Your great patch is much appreciated. We are considering to apply your patch into the SoftEther VPN main tree.

SoftEther VPN Patch Acceptance Policy:
http://www.softether.org/5-download/src/9.patch

You have two options which are described on the above policy.
Could you please choose either option 1 or 2, and specify it clearly on the reply?

I am ok with option 1
-

### CODE REVIEW

**Comment by fenice2** [comment]
You just need to pot whether you prefer option 1 or 2. :)

**Comment by moatazelmasry2** [comment]
Small tip. Everyone usually selects Option 1 :)

**Comment by youxiaojie** [comment]
I AM OK with option 1

**Comment by youxiaojie** [comment]
when will my code be merged into upstream?

 
------------------ Original ------------------

**Comment by paskal** [comment]
@moatazelmasry2, could you please take a look if it could be merged now?

**Comment by youxiaojie** [comment]
@paskal please check my commits and give me some advice.

**Comment by paskal** [comment]
@youxiaojie I am not a maintainer, it's better to ask @moatazelmasry2 opinion. However, from my POV it's quite complex - at least you should write a summary, what your merge request is about.

**Comment by youxiaojie** [comment]
@moatazelmasry2 , could you give me some advice about my commit?

**Comment by moatazelmasry2** [comment]
Hi @youxiaojie thanks for the hard work. I'm still in christmas/new year's mode. and will  be back on 8th of january. Will try to get it done then. Promise :)

File: README.md [hunk 1]
```diff
@@ -140,17 +140,19 @@ from GitHub. You may make your own fork project from our project.
 ### 1. Debian/Ubuntu
 
 - gcc
-- libncurses-dev
-- libreadline-dev
 - make
-- openssl-dev
+- libreadline-dev
```
Comment (moatazelmasry2) [inline README.md]: Why did you move those three lines to the bottom of the list? It was by intention that the list is alphabetically sorted

File: README.md [hunk 2]
```diff
@@ -140,17 +140,19 @@ from GitHub. You may make your own fork project from our project.
...
+- zlib1g-dev
 
 ### 2. Redhat/CentOS
 - gcc
-- openssl-devel
 - make
-- ncurses-devel
 - readline-devel
+- openssl-devel
+- ncurses-devel
```
Comment (moatazelmasry2) [inline README.md]: List should be alphabetically sorted. And where is the dependancy on zlib for Redhat?

File: centos/SPECS/softethervpn.spec [hunk 3]
```diff
@@ -1,7 +1,7 @@
 %define majorversion 4
-%define minorversion 19
-%define buildversion 9582
-%define dateversion 2015.10.06
+%define minorversion 23
```
Comment (moatazelmasry2) [inline centos/SPECS/softethervpn.spec]: Why 23? 24 has already been released

File: centos/SPECS/softethervpn.spec [hunk 4]
```diff
@@ -1,7 +1,7 @@
...
+%define buildversion 9647
```
Comment (moatazelmasry2) [inline centos/SPECS/softethervpn.spec]: 9651 has already been released

File: centos/SPECS/softethervpn.spec [hunk 5]
```diff
@@ -87,6 +87,25 @@ if [ $1 -eq 0 ]; then
 fi
 
 %changelog
+* Wed Dec 18 2017 You Xiaojie <yxj790222@163.com> - 4.23.9647-1
+- Upgraded OpenSSL to 1.0.2l.
```
Comment (moatazelmasry2) [inline centos/SPECS/softethervpn.spec]: Question: How did you find this list?

File: src/Cedar/BridgeUnix.h [hunk 6]
```diff
@@ -186,6 +186,7 @@ struct CAPTUREBLOCK{
 
 
 // Function prototype
+void FreeTap(VLAN *v);
```
Comment (moatazelmasry2) [inline src/Cedar/BridgeUnix.h]: This function is implemented at src/Cedar/VLanUnix.c line 388. So the definition should also be under: src/Cedar/VLanUnix.h

File: src/Cedar/Console.c [hunk 7]
```diff
@@ -2048,7 +2048,13 @@ bool PasswordPrompt(char *password, UINT size)
 		else if (c == 0xE0)
 		{
 			// Read one more character
+
+#ifdef	OS_WIN32
 			c = getch();
+#else	// OS_WIN32
```
Comment (moatazelmasry2) [inline src/Cedar/Console.c]: If OS_WIN32 then else OS_WIN32 ?? The comment after "else" does not make sense

File: src/Cedar/Console.c [hunk 8]
```diff
@@ -2048,7 +2048,13 @@ bool PasswordPrompt(char *password, UINT size)
 		else if (c == 0xE0)
 		{
 			// Read one more character
+
+#ifdef	OS_WIN32
 			c = getch();
+#else	// OS_WIN32
+			c = getc(stdin);
```
Comment (moatazelmasry2) [inline src/Cedar/Console.c]: Which problem does this line solve???

**Comment by youxiaojie** [comment]
在2018年01月12 05时36分, "Moataz Elmasry"<notifications@github.com>写道:



@moatazelmasry2 requested changes on this pull request.

Thanks for the PR. I agree with @paskal that this PR is quite complex and does too many things.  Please consider my comments on the single lines and divide this PR at least into 3 Pull requests:
1- for changes in README.md
2- for changes in the changelog
3- for the FreeTap() addition

And if you can explain your change in src/Cedar/Console.c, then you will need a fourth pull request. Please remember with every Pull Request to add enough comment to explain what you have done.

Thank you

In README.md:

>  - make
-- openssl-dev
+- libreadline-dev

Why did you move those three lines to the bottom of the list? It was by intention that the list is alphabetically sorted

ok.I will correct.this sort is the package I install for a newly insalled system, if compile and link error, I will install by the feedback message. all by my test.


In README.md:

>  - readline-devel
+- openssl-devel
+- ncurses-devel

List should be alphabetically sorted. And where is the dependancy on zlib for Redhat?

This is your own. not changed. conference by others.


In centos/SPECS/softethervpn.spec:

> @@ -1,7 +1,7 @@
 %define majorversion 4
-%define minorversion 19
-%define buildversion 9582
-%define dateversion 2015.10.06
+%define minorversion 23

Why 23? 24 has already been released

maybe I built weeks ago and not test again immidiately


In centos/SPECS/softethervpn.spec:

> @@ -1,7 +1,7 @@
 %define majorversion 4
-%define minorversion 19
-%define buildversion 9582
-%define dateversion 2015.10.06
+%define minorversion 23
+%define buildversion 9647

9651 has already been released




maybe I built weeks ago and not test again immidiately

In centos/SPECS/softethervpn.spec:

> @@ -87,6 +87,25 @@ if [ $1 -eq 0 ]; then
 fi
 
 %changelog
+* Wed Dec 18 2017 You Xiaojie <yxj790222@163.com> - 4.23.9647-1
+- Upgraded OpenSSL to 1.0.2l.

Question: How did you find this list?

from your Changelog files on project home page

In src/Cedar/BridgeUnix.h:

> @@ -186,6 +186,7 @@ struct CAPTUREBLOCK{
 
 
 // Function prototype
+void FreeTap(VLAN *v);

This function is implemented at src/Cedar/VLanUnix.c line 388. So the definition should also be under: src/Cedar/VLanUnix.h

but the question is in VLanUnix.c, the compiler find a undeclared function which defined in other file infact. so it will warning "implicit declaration". Therefore I add a declaration before using void FreeTap(VLAN *).

In src/Cedar/Console.c:

>  			c = getch();
+#else	// OS_WIN32

If OS_WIN32 then else OS_WIN32 ?? The comment after "else" does not make sense

this is I got context for respect original programmer.do nothing. I think he/she is correct, which means following is not for os_win32 and for easily find win32 crossponding.


In src/Cedar/Console.c:

>  			c = getch();
+#else	// OS_WIN32
+			c = getc(stdin);

Which problem does this line solve???

I find the C lib refenrence, getch() is not a stdio.h standard function, also got "implicit declaration" error.


. On the other hand the original auther using a conditional compile to solve it in the upper context. so I copy and paste

File: README.md (hunk 1 above)
Comment (youxiaojie) [reply README.md]: ok.I will correct.this sort is the package I install for a newly insalled system, if compile and link error, I will install by the feedback message. all by my test.

File: README.md (hunk 2 above)
Comment (youxiaojie) [reply README.md]: This is your own. not changed. conference by others.if needed I will test.

File: centos/SPECS/softethervpn.spec (hunk 3 above)
Comment (youxiaojie) [reply centos/SPECS/softethervpn.spec]: according to your changelog file on the root directory 
https://github.com/SoftEtherVPN/SoftEtherVPN/blob/master/ChangeLog

File: centos/SPECS/softethervpn.spec (hunk 4 above)
Comment (youxiaojie) [reply centos/SPECS/softethervpn.spec]: https://github.com/SoftEtherVPN/SoftEtherVPN/blob/master/ChangeLog
it is said 9647:(

File: centos/SPECS/softethervpn.spec (hunk 5 above)
Comment (youxiaojie) [reply centos/SPECS/softethervpn.spec]: https://github.com/SoftEtherVPN/SoftEtherVPN/blob/master/ChangeLog
and which  version I would using? the version shown by  running binary or this file?

File: src/Cedar/Console.c (hunk 8 above)
Comment (youxiaojie) [reply src/Cedar/Console.c]: I find the C lib reference, getch() is not a stdio.h standard function, also got "implicit declaration" error.

. On the other hand the original author using a conditional compile to solve it in the upper context. so I copy and paste

please see the old file line 2019~2023

File: src/Cedar/BridgeUnix.h (hunk 6 above)
Comment (youxiaojie) [reply src/Cedar/BridgeUnix.h]: but the question is VLanUnix.c, the compiler find a undeclared function which defined in other file in fact. so it will warning "implicit declaration". Therefore I add a declaration before using void FreeTap(VLAN *).
is it necessary? you are export I following your advice.

see code:
https://github.com/SoftEtherVPN/SoftEtherVPN/blob/master/src/Cedar/BridgeUnix.c#L1570
FreeTap was define in other files. Books say  that if call the function defined in another file we should add function prototype before calling.

**Comment by moatazelmasry2** [comment]
@youxiaojie yes please try to make it clear. Best to separe this pull request into multiple PRs where each one does exactly one thing only. Thanks
